File: src/main/java/com/bftcom/mediastorage/model/entity/Category.java
package com.bftcom.mediastorage.model.entity;

import lombok.AllArgsConstructor;
import lombok.Data;
import lombok.EqualsAndHashCode;
import lombok.NoArgsConstructor;

import javax.persistence.*;
import javax.validation.constraints.NotBlank;
import javax.validation.constraints.Size;
import java.util.HashSet;
import java.util.Set;

@Data
@NoArgsConstructor
@AllArgsConstructor
@EqualsAndHashCode(callSuper = true)
@Entity
@Table(name = "category")
public class Category extends BaseEntity {

    @NotBlank
    @Size(max = 200)
    @Column(name = "name")
    private String name;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "parent_category_id")
    private Category parentCategory;

    @ManyToMany
    @JoinTable(
            name = "user_role",
            joinColumns = @JoinColumn(name = "user_id"),
            inverseJoinColumns = @JoinColumn(name = "role_id")
    )
    private Set<Role> roles = new HashSet<>();

    public Category(Long id, String name, Category parentCategory) {
        super(id);
        this.name = name;
        this.parentCategory = parentCategory;
    }
}

File: src/main/js/models/Category.js
class Category {
    constructor() {
        this.id = null;
        this.name = "";
        this.parentCategoryId = null;
    }
}

export const CategoryBuilder = {
    getDefault: () => new Category(),

    buildByParams: (id, name, parentCategoryId) => {
        let category = new Category();
        category.id = id;
        category.name = name;
        category.parentCategoryId = parentCategoryId;
        return category;
    },

    buildByData: (data) => {
        return CategoryBuilder.buildByParams(
            data.id,
            data.name,
            data.parentCategoryId
        )
    }
}